src/gleam/dynamic.gleam

```diff
@@ -474,20 +476,77 @@ if javascript {
 ///
 pub fn field(named name: a, of inner_type: Decoder(t)) -> Decoder(t) {
   fn(value) {
-    value
-    |> decode_field(name)
-    |> result.try(inner_type)
-    |> map_errors(push_path(_, name))
+    case decode_field(value, name) {
+      Error(not_a_map_errors) -> Error(not_a_map_errors)
+      Ok(dynamic_field_result) ->
+        dynamic_field_result
+        |> result.try(inner_type)
+        |> map_errors(push_path(_, name))
+    }
+  }
+}
+
+/// Checks to see if a `Dynamic` value is a map with a specific field.
+/// If the map does not have the specified field, returns an `Ok(None)` instead of failing; otherwise,
+/// returns the decoded field wrapped in `Some(_)`.
+///
+/// ## Examples
+///
+/// ```gleam
+/// > import gleam/map
+/// > map.new()
+/// > |> map.insert("Hello", "World")
+/// > |> from
+/// > |> field(named: "Hello", of: string)
+/// Ok(Some("World"))
+/// ```
+///
+/// ```gleam
+/// > import gleam/map
+/// > map.new()
+/// > |> from
+/// > |> field(named: "Hello", of: string)
+/// Ok(None)
+/// ```
+///
+/// ```gleam
+/// > from(123)
+/// > |> field("Hello", string)
+/// Error([DecodeError(expected: "Map", found: "Int", path: [])])
+/// ```
+///
+pub fn optional_field(
+  named name: a,
+  of inner_type: Decoder(t),
+) -> Decoder(Option(t)) {
+  fn(value) {
+    case decode_field(value, name) {
+      Error(not_a_map_errors) -> Error(not_a_map_errors)
+      Ok(dynamic_field_result) ->
+        case dynamic_field_result {
+          Error(_) -> Ok(option.None)
+          Ok(dynamic_field) ->
+            dynamic_field
+            |> decode_optional(inner_type)
+            |> map_errors(push_path(_, name))
+        }
+    }
   }
 }
 
 if erlang {
-  external fn decode_field(Dynamic, name) -> Result(Dynamic, DecodeErrors) =
+  external fn decode_field(
+    Dynamic,
+    name,
+  ) -> Result(Result(Dynamic, DecodeErrors), DecodeErrors) =
```

Comment:
How about this returns `Result(Option(Dynamic), DecodeErrors)`? A slightly more approachable signature.

Reply:
I thought about it to make it a little easier to read; however, this way one loses the ability to provide detailed error messages in case the inner decoder fails. Consider this example: 
```gleam
  map.new()
  |> map.insert("ok", 3)
  |> dynamic.from
  |> dynamic.field("ok", dynamic.string)
```
The result of a call to `decode_field` would be an `Ok(None)` (the `Dynamic` is actually a `map` but the field decoding failed), discarding the `Result` I would have no way to know what to put in the `expected` and `found` fields of the decoding error: `DecodingError(expected: "?", found: "?", path: ["ok"])`